04_basic_struct/lds_module.c

```diff
@@ -0,0 +1,216 @@
+// SPDX-License-Identifier: GPL-2.0 OR BSD-3-Clause
+
+#include <linux/init.h>
+#include <linux/module.h>
+#include <linux/kernel.h>
+#include <linux/printk.h>
+
+#include <linux/sysfs.h>
+#include <linux/device.h>
+
+#include <linux/types.h>
+#include <linux/list.h>
+#include <linux/slab.h>
+
+//#define DEBUG
+
+MODULE_AUTHOR("Andrii Synenko");
+MODULE_DESCRIPTION("List data module for Linux Kernel ProCamp");
+MODULE_LICENSE("Dual BSD/GPL");
+
+struct data_node {
+	struct list_head data_list;
+	size_t data_size;
+	const char *data_ptr;
+};
+
+struct accum_data {
+	struct list_head accum_list;
+	size_t accum_size;
+} gl_accum;
+
+static int accum_entry_add(struct accum_data *accum, const char *new_data_prt,
+						  size_t new_data_size)
+{
+	struct data_node *entry;
+
+	if ((accum == NULL) || (new_data_prt == NULL)) {
+		pr_err("Linked list of input doesn't exist!\n");
+		return -EINVAL;
+	}
+
+	entry = kmalloc(sizeof(*entry), GFP_KERNEL);
+	if (entry == NULL)
+		return -ENOMEM;
+
+	entry->data_ptr = kstrdup(new_data_prt, GFP_KERNEL);
+
+	if (entry->data_ptr == NULL)
+		return -ENOMEM;
+
+	entry->data_size = new_data_size;
+	accum->accum_size += entry->data_size;
+	list_add_tail(&entry->data_list, &accum->accum_list);
+
+	return 0;
+}
+
+static int accum_free_from_head(struct accum_data *aaccum, size_t req_space)
+{
+	struct data_node *entry = NULL;
+
+	if (aaccum == NULL) {
+		pr_err("Linked list doesn't exist!\n");
+		return -EINVAL;
+	}
+
+	while ((aaccum->accum_size + req_space) > PAGE_SIZE) {
+
+		if (list_empty(&aaccum->accum_list) == 1)
+			return -EINVAL;
+
+		entry = list_first_entry(&aaccum->accum_list, struct data_node,
+								 data_list);
+
+		list_del(&entry->data_list);
+		kfree(entry->data_ptr);
+		aaccum->accum_size -= entry->data_size;
+		kfree(entry);
+	}
+
+	return 0;
+}
+
+static int accum_print(struct accum_data *aaccum, char *buf)
+{
+	size_t list_size = 0;
+	int res = 0;
+	int cursor = 0;
+
+	struct data_node *entry = NULL;
+
+	if ((aaccum == NULL) || (buf == NULL)) {
+		pr_err("Linked list or output buffer doesn't exist!\n");
+		return -EINVAL;
+	}
+
+	list_for_each_entry(entry, &aaccum->accum_list, data_list) {
+		res = sprintf(&buf[cursor], "%s\n", entry->data_ptr);
+		if (res > 0)
+			cursor += res;
+		else {
+			pr_err("Failed to write output.\n");
+			return -EIO;
+		}
+
+		list_size++;
+	}
+
+	return cursor;
+}
+
+static int accum_free(struct accum_data *aaccum)
+{
+	struct data_node *entry = NULL,
+					 *tmp = NULL;
+
+	if (aaccum == NULL) {
+		pr_err("Linked list doesn't exist!\n");
+		return -EINVAL;
+	}
+
+	list_for_each_entry_safe(entry, tmp, &aaccum->accum_list, data_list) {
+
+		list_del(&entry->data_list);
+		kfree(entry->data_ptr);
+		aaccum->accum_size -= entry->data_size;
+		kfree(entry);
```

Comment:
Add a helper function to use
```c
  accum_size -= entry_del(entry);
```
instead of these four lines and use it here and in the `accum_free_from_head()` finction.

Reply:
Helper function added